typesetters/base.lua

```diff
@@ -46,16 +46,35 @@ end
 -- @param frame A initial frame to attach the typesetter to.
 function typesetter:_init (frame)
    SU._avoid_base_class_use(self)
+   -- TODO: make class first arg of typesetter init, ditch globals hack
+   self.class = SILE.documentState.documentClass
```

Comment:
Why would the typesetter(s) have to know anything about the document class, by the way?

Reply:
Among other reasons, because it needs access to call the `newPar()`, `endPar()`, `newPage()`, and `endPage()` functions that are defined in the class.

...and I'm working towards removing all the global state stuff (`SILE.scratch`, `SILE.documentState`, etc.).

Reply:
To save you the trouble of searching, see 2fab290d2. Those are example of accessing global state variables that I'm trying to eliminate. Not having that kind of nonsense will make actually using and extending SILE **much much** easier because each piece should actually be modular with known inputs and outputs and no side effects.

Reply:
(endPage/newPage are _page builder_ concepts = inherently not tied to the typesetter. Just saying, _this is nonsense in a layered architecture._)

Reply:
Agreed. Lots of stuff is in the wrong layers.

I'm leaning towards giving *all* module types some kind of access to navigate the entire tree of modules that make up the active SILE session rather than a minimalist approach of only having what our current implementation uses. Right now this is already the case because we just reach into global state when we want something, but in the process of removing that it isn't always clear what *may be* needed by some alternative implementation. I'm currently proposing all modules have some link into the current process internals: not necessarily as a list of all current modules, but as it's most logical parent and most relevant children.